src/Lean/Meta/Tactic/Simp/Main.lean

```diff
@@ -458,11 +458,13 @@ private partial def dsimpImpl (e : Expr) : SimpM Expr := do
   let cfg ← getConfig
   unless cfg.dsimp do
     return e
+  let dsimpCache ← modifyGet fun s => (s.dsimpCache, { s with dsimpCache := {} })
```

Comment:
```suggestion
  -- replace the cache in the state to ensure the old cache is used linearly.
  let dsimpCache ← modifyGet fun s => (s.dsimpCache, { s with dsimpCache := {} })
```
I'd be tempted to move this to after line 464, just to make it super clear that nothing is going to touch the cache after it is cleared.